client/hotrod-client/src/test/java/org/infinispan/client/hotrod/ExternalAddressRoutingTest.java

```diff
@@ -0,0 +1,194 @@
+      realAddressesToProxy.put(new InetSocketAddress(server2.getHost(), server2.getPort()), new InetSocketAddress("localhost", PROXY_TO_SERVER_2_PORT));
+
+      org.infinispan.client.hotrod.configuration.ConfigurationBuilder clientBuilder = new org.infinispan.client.hotrod.configuration.ConfigurationBuilder();
+      // We add only one server here. Once the HR client initiates connection, it should have two addresses in the connection pool.
+      clientBuilder
+            .addServer()
+               .host("localhost").port(PROXY_TO_SERVER_2_PORT)
+            .addressMapping(LoopbackAddressMapper.class);
+      rcm = new RemoteCacheManager(clientBuilder.build());
+   }
+
+   @AfterClass
+   @Override
+   protected void destroy() {
+      killServers(server1, server2);
+      proxyToServer1.shutdown();
+      proxyToServer2.shutdown();
+      super.destroy();
+   }
+
+   private HotRodServer addHotRodServer() {
+      EmbeddedCacheManager cm = addClusterEnabledCacheManager(defaultBuilder);
+      cm.defineConfiguration(DIST_ONE_CACHE_NAME, distOneBuilder.build());
+      cm.defineConfiguration(DIST_TWO_CACHE_NAME, distTwoBuilder.build());
+      HotRodServer server = HotRodClientTestingUtil.startHotRodServer(cm);
+      addr2hrServer.put(new InetSocketAddress(server.getHost(), server.getPort()), server);
+      return server;
+   }
+
+   public void testRoutingWithExternalAddresses() {
+      addInterceptors(DIST_ONE_CACHE_NAME);
+      addInterceptors(DIST_TWO_CACHE_NAME);
+
+      byte[] keyDistOne = getKeyForServer(server1, DIST_ONE_CACHE_NAME);
+      byte[] keyDistTwo = getKeyForServer(server2, DIST_TWO_CACHE_NAME);
+
+      assertRequestRouting(keyDistOne, DIST_ONE_CACHE_NAME, server1);
+      assertRequestRouting(keyDistTwo, DIST_TWO_CACHE_NAME, server2);
+      assertProperServerAddressInTopology(DIST_ONE_CACHE_NAME, PROXY_TO_SERVER_1_PORT);
+      assertProperServerAddressInTopology(DIST_TWO_CACHE_NAME, PROXY_TO_SERVER_2_PORT);
+
+   }
+
+   private void assertProperServerAddressInTopology(String cacheName, int serverPort) {
+      CacheTopologyInfo cache1Topology = rcm.getTransportFactory().getCacheTopologyInfo(cacheName.getBytes());
+      for (SocketAddress serverAddress : cache1Topology.getSegmentsPerServer().keySet()) {
+         int port = ((InetSocketAddress) serverAddress).getPort();
+         assertEquals(serverPort, port);
+      }
+   }
+
+   private void assertRequestRouting(byte[] key, String cacheName, HotRodServer server) {
+      RemoteCache<Object, Object> rcOne = rcm.getCache(cacheName);
+      InetSocketAddress serverAddress = new InetSocketAddress(server.getHost(), server.getPort());
+      for (int i = 0; i < 2; i++) {
+         log.infof("Routing put test for key %s", Util.printArray(key, false));
+         rcOne.put(key, "value");
+         assertServerHit(serverAddress, cacheName, i + 1);
+      }
+   }
+
+   byte[] getKeyForServer(HotRodServer primaryOwner, String cacheName) {
```

Comment:
We should avoid code duplication, `HotRodClientTestingUtil` contains this method.